### DISCUSSION TITLE
useResult deprecation question

### DISCUSSION THREAD

**Discussion by wreality** [Akryum/vue-apollo#1354] [Q&A] [answered]
Is it safe to assume that with getResult being [removed](https://github.com/vuejs/apollo/commit/2e65263329fa8c9d30cd681778bf7a8fcc22bfed)  from the documentation (rather than a deprecation note added to the documentation) that the intention is that it will be removed for the 4.0 release ?

Just figuring out where it needs to live in our project's priorities.

**Comment by Luferov** [comment]
And in general, why useResult is deprecated? What is the reason? Because this function is very usefull.

**Comment by Akryum** [comment, marked as answer]
@wreality Yes it will be removed before 4.0

@Luferov It's not that useful anymore now that optional chaining and nullish coalescing operators are available in standard JavaScript. So it makes sense to remove it to cleanup the API a bit.

**Reply by Luferov** [reply]
Yep, I agree, but if use one endpoint from library at project, what is good practice, use result is good solution for different queries. Because than I need always use result.data.foo, result.data.bar instead one solution useResult.

**Reply by Akryum** [reply]
You can write a function to do that:

```js
export function autoPick (value) {
  const keys = Object.keys(value)
  if (keys.length === 1) {
    // Automatically take the only key in result data
    return value[keys[0]]
  } else {
    // Return entire result data
    return value
  }
}
```

And then use it like this:

```js
const items = computed(() => autoPick(result.value))
```

Or write a composable version with the 'computed' part inside.

**Reply by Akryum** [reply]
I don't feel maintaining code and documentation for this is worthwhile in the library.

**Reply by Luferov** [reply]
Ok, thanks 🙏

**Comment by sschneider-ihre-pvs** [comment]
This might get a nightmare if you have https://eslint.vuejs.org/rules/no-side-effects-in-computed-properties.html on and try to swap out `useResult` for a computed.

**Reply by Akryum** [reply]
Why is that?

**Reply by sschneider-ihre-pvs** [reply]
If you had side effects in the callback function before and try to just swap the useResult for a computed you will get an error.

**Reply by sschneider-ihre-pvs** [reply]
Quick example
```
useResult(result, undefined, (data) => {
   someStatus.value = true
   return data;
}
```
```
computed(() => {
   if (result) {
     someStatus.value = true;
     return result;
   }

   return undefined;
}
```

**Reply by Akryum** [reply]
I guess it's a plus now that this bad code will get catched by ESLint :)

**Reply by sschneider-ihre-pvs** [reply]
Well, yes, but it will also result in a lot of head scratching, since `computed` in this case might not be the right choice but maybe `watch` might help, just in reference to the deprecation text.

**Reply by Akryum** [reply]
Well, the third parameter of `useResult` was never supposed to contain side-effects in the first place.

**Reply by sschneider-ihre-pvs** [reply]
Well I would try and point at the documentation, but `useResult` is already gone :D

**Reply by sschneider-ihre-pvs** [reply]
> Well, the third parameter of `useResult` was never supposed to contain side-effects in the first place.

What would be the correct place for side-effects?

**Reply by Akryum** [reply]
A watcher or `onResult`

**Comment by shtse8** [comment]
Any reason why undefined is trigger when loading? It is nightmare because all others using the result as a variable for their query will also be triggered when one is loading. How can I bypass the undefined value when using it as a variable of query but not for displaying?
